fix: Add Loader on POST Actions in Discussions

maheshchathani opened the pull request:
Loader is shown during each POST action
Button becomes disabled during request
User receives clear success or failure feedback

maheshchathani committed:
fix: Add Loader on POST Actions in Discussions

maheshchathani committed:
fix: Add Loader on POST Actions in Discussions

HamzaIsrar12 commented on discussion/src/main/java/org/openedx/discussion/presentation/comments/DiscussionCommentsViewModel.kt:
```diff
@@ -66,7 +66,10 @@ class DiscussionCommentsViewModel(
+    private val _isLoading = MutableLiveData<Boolean>()
+    val isLoading: LiveData<Boolean>
+        get() = _isLoading
+    //private var isLoading = false
```
Hi Mahesh, we should avoid removing or commenting out the original `isLoading` variable. It ensures that the next page isn’t requested while the current one is still loading. If you feel the name `isLoading` is ambiguous, you can go ahead and rename it.

For the loader of the new comment, I’d recommend creating a separate `MutableStateFlow` so both states remain independent. The same approach should also be applied in `DiscussionResponsesViewModel.kt`.

Why `MutableFlowState` over `LiveData`? Please [read](https://github.com/edx/edx-mobile-marketplace-android/blob/fix/navigation/docs/0001-strategy-for-data-streams.rst) this for it. 😉

maheshchathani replied:
Hi Hamza, as suggested, reverted isLoading variable to previous state. 
added new variable "showProgress" using MutableStateFlow.

HamzaIsrar12 commented on discussion/src/main/java/org/openedx/discussion/presentation/comments/DiscussionCommentsViewModel.kt:
```diff
@@ -315,6 +319,8 @@ class DiscussionCommentsViewModel(
                 notifier.send(DiscussionCommentAdded())
             } catch (e: Exception) {
                 handleException(e)
+            }finally {
```
Please reformat the code. 😄

maheshchathani replied:
Done, Code reformatted.

HamzaIsrar12 commented on discussion/src/main/java/org/openedx/discussion/presentation/comments/DiscussionCommentsFragment.kt:
```diff
@@ -256,6 +260,7 @@ private fun DiscussionCommentsScreen(
     uiMessage: UIMessage?,
     title: String,
     canLoadMore: Boolean,
+    isloading: Boolean,
```
Android Studio flags typos in the top-left corner of the editor. Please resolve these before pushing. In this case, the variable should be `isLoading` (camelCase). The same change should also be applied in `DiscussionResponsesFragment.kt`

(Based on another comment, this variable will be renamed to something else.)

maheshchathani replied:
Followed camel case for newly added variable.

HamzaIsrar12 commented on discussion/src/main/java/org/openedx/discussion/presentation/comments/DiscussionCommentsFragment.kt:
```diff
@@ -310,7 +315,16 @@ private fun DiscussionCommentsScreen(
             )
         }
-
+        if (!uiState.equals(DiscussionResponsesUIState.Loading) && isloading){
```
Logical issue: we’re comparing a `DiscussionCommentsUIState` to a `DiscussionResponsesUIState` value. Also, we don’t need this check here; `isLoading` is sufficient to control this UI.

maheshchathani replied:
Made changes as per the comments.

HamzaIsrar12 commented on discussion/src/main/java/org/openedx/discussion/presentation/comments/DiscussionCommentsFragment.kt:
```diff
@@ -310,7 +315,16 @@ private fun DiscussionCommentsScreen(
         }
-
+        if (!uiState.equals(DiscussionResponsesUIState.Loading) && isloading){
+            Dialog(onDismissRequest = { /* Do nothing to prevent dismiss */ }) {
+                Box(
+                    Modifier
+                        .fillMaxSize(), contentAlignment = Alignment.Center
+                ) {
+                    CircularProgressIndicator(color = MaterialTheme.appColors.primary)
+                }
+            }
+        }
```
Instead of a full-screen loading dialog, let’s show an inline spinner in the send button (consistent with the rest of the app) so the UI remains interactive and non-blocking.

```
    if (!isloading) {
        Icon(
            modifier = Modifier.padding(7.dp),
            painter = painterResource(id = R.drawable.discussion_ic_send),
            contentDescription = stringResource(id = R.string.discussion_add_response),
            tint = MaterialTheme.appColors.primaryButtonText
        )
    } else {
        CircularProgressIndicator(
            modifier = Modifier.padding(7.dp),
            color = MaterialTheme.appColors.primary,
        )
    }
```

We may need to update the `sendButtonAlpha` as well: 
```
val sendButtonAlpha = if (responseValue.isEmpty() && !isLoading) 0.3f else 1f
```

The same approach should also be applied in `DiscussionResponsesFragment.kt`

maheshchathani replied:
for this, as suggested, removed full-screen loader and added inline button loader for send buttons in both comments and responses.
But the thing is, added loader not only in place of icon, but added in place entire box to skip the background colour to the loader. so here I guess button alpha not required to modify.

HamzaIsrar12 replied:
Perfect! We can also go live with this approach, it just needs a small tweak. You’ll notice that the `CircularProgressIndicator` takes up more space than the send button, causing a slight UI shift.


https://github.com/user-attachments/assets/8de879ea-1cbb-47cc-a960-5b31c5293723

maheshchathani replied:
Hi Hamza, as suggested, UI shift/glitch issue resolved. Please find the attached record and review it again.

https://github.com/user-attachments/assets/1d7ee7a1-8a54-495e-9677-f358259d4322

HamzaIsrar12 reviewed (changes requested):
I have added some comments. 🙌🏻 

---

Some OpenEdx and the Team's convention:

- Attach the ticket here in the PR description so it's easier to navigate to it.
- Moving forward, please add the ticket number in the commit body.

For reference, you can check the PR description and commits here: https://github.com/edx/edx-mobile-marketplace-android/pull/161

omerhabib26 committed:
refactor: Code Cleanup & Imports Improvement (#164)

- Prioritise module-specific R import over core
- Code cleanup using the Android Studio tool

Fix: LEARNER-10647

maheshchathani committed:
refactor: Add Loader on POST Actions in Discussions

maheshchathani committed:
refactor: Add Loader on POST Actions in Discussions

maheshchathani commented:
Hi Hamza, as per the mentioned comments, made changes and pushed the code into same PR.
Please review it again. If anything required. please let me know.

Ticket : [LEARNER-10727](https://2u-internal.atlassian.net/browse/LEARNER-10727)

HamzaIsrar12 committed:
build: Release v6.2.2 (#171)

Fixes: LEARNER-10670

maheshchathani committed:
refactor: Add Loader on POST Actions in Discussions

maheshchathani committed:
refactor: Add Loader on POST Actions in Discussions

kalalmounesh committed:
feat: Refresh Discussions UI and Scroll to Response After Posting (#170)

Fixes: LEARNER-10728

maheshchathani committed:
Merge remote-tracking branch 'origin/mahesh/LEARNER-10727' into mahesh/LEARNER-10727
